pageserver/src/layered_repository.rs

```diff
@@ -289,131 +304,203 @@ impl Repository for LayeredRepository {
     // Sync task is enqueued and can error and be rescheduled, so some significant time may pass between the events.
     //
     /// Reacts on the timeline sync state change, changing pageserver's memory state for this timeline (unload or load of the timeline files).
-    fn set_timeline_state(
-        &self,
-        timeline_id: ZTimelineId,
-        new_state: TimelineSyncState,
-    ) -> Result<()> {
+    fn set_timeline_state(&self, timeline_id: ZTimelineId, new_state: TimelineState) -> Result<()> {
         let mut timelines_accessor = self.timelines.lock().unwrap();
 
-        match new_state {
-            TimelineSyncState::Ready(_) => {
+        let timeline_to_shutdown = match new_state {
+            TimelineState::Local(local_state) => {
+                // TODO (current) rebase this first and then handle all the options
+
+                // TODO do not load timeline forcibly, layer map loading can take significant time,
+                // so it is benefitial to just mark timeline as existing locally without loading it.
+                // (now we do that under timelines mutex which is bad by itself too)
                 let reloaded_timeline =
-                    self.init_local_timeline(timeline_id, &mut timelines_accessor)?;
-                timelines_accessor
-                    .insert(timeline_id, LayeredTimelineEntry::Local(reloaded_timeline));
+                    self.load_local_timeline(timeline_id, &mut timelines_accessor)?;
+                timelines_accessor.insert(
+                    timeline_id,
+                    LayeredTimelineEntry::Local(LocalLayeredTimelineEntry::Loaded(
+                        reloaded_timeline,
+                    )),
+                );
                 None
             }
-            TimelineSyncState::Evicted(_) => timelines_accessor.remove(&timeline_id),
-            TimelineSyncState::AwaitsDownload(disk_consistent_lsn)
-            | TimelineSyncState::CloudOnly(disk_consistent_lsn) => timelines_accessor.insert(
-                timeline_id,
-                LayeredTimelineEntry::Remote {
-                    id: timeline_id,
-                    disk_consistent_lsn,
-                },
-            ),
+            TimelineState::Evicted(_) => timelines_accessor.remove(&timeline_id), // FIXME Should we shutdown the timeline?
```

Comment:
Before that, we at least disabled its S3 sync capabilities at this moment.
That was removed in https://github.com/zenithdb/zenith/pull/1108#discussion_r783391549 , so might not be need with the approach?

Reply:
Evicted state is removed for now, will see how it works out in a follow up